mindsdb/integrations/handlers/dspy_handler/dspy_handler.py

```diff
@@ -0,0 +1,211 @@
+        if args['model_name'] in OPEN_AI_CHAT_MODELS:
+            return 'openai'
+        if args['model_name'] in OLLAMA_CHAT_MODELS:
+            return 'ollama'
+        raise ValueError(f"Invalid model name. Please define provider")
+
+    def _get_embedding_model_provider(self, args: Dict) -> str:
+        if 'embedding_model_provider' in args:
+            return args['embedding_model_provider']
+        if 'embedding_model_provider' not in args:
+            logger.warning('No embedding model provider specified. trying to use llm provider.')
+            return args.get('embedding_model_provider', self._get_llm_provider(args))
+        raise ValueError(f"Invalid model name. Please define provider")
+
+    def create(self, target: str, df: Optional[pd.DataFrame] = None, args: Dict = None, **kwargs):
+
+        args = args['using']
+        args['target'] = target
+        args['model_name'] = args.get('model_name', DEFAULT_MODEL_NAME)
+        args['provider'] = args.get('provider', self._get_llm_provider(args))
+        args['embedding_model_provider'] = args.get('embedding_model', self._get_embedding_model_provider(args))
+        if args.get('mode') == 'retrieval':
+            # use default prompt template for retrieval i.e. RAG if not provided
+            if "prompt_template" not in args:
+                args["prompt_template"] = DEFAULT_RAG_PROMPT_TEMPLATE
+
+        self.model_storage.json_set('args', args)
+        self.model_storage.file_set("cold_start_df", dill.dumps(df.to_dict()))
+            # TODO: temporal workaround: serialize df and args, instead. And recreate chain (with training) every inference call.
+            # ideally, we serialize the chain itself to avoid duplicate training.
+
+    def predict(self, df: pd.DataFrame, args: Dict=None) -> pd.DataFrame:
+        """
+        Dispatch is performed depending on the underlying model type. Currently, only the default text completion
+        is supported.
+        """
+        pred_args = args['predict_params'] if args else {}
+        args = self.model_storage.json_get('args')
+        if 'prompt_template' not in args and 'prompt_template' not in pred_args:
+            raise ValueError(f"This model expects a `prompt_template`, please provide one.")
+        # Back compatibility for old models
+        args['provider'] = args.get('provider', self._get_llm_provider(args))
+        args['embedding_model_provider'] = args.get('embedding_model', self._get_embedding_model_provider(args))
+
+        # retrives llm and pass it around as context
+        model = args.get('model_name')
+        api_key = get_api_key('openai', args, self.engine_storage, strict=False)
+        # llm = dspy.OpenAI(model=model, api_key=api_key, api_base = 'https://llm.mdb.ai')
+        llm = dspy.OpenAI(model=model, api_key=api_key)
+
+        df = df.reset_index(drop=True)
+        if len(self.llm_data_controller.list_all_llm_data(1)) > 100:
+            for i in range(15):
```

Comment:
I was using this as a way to make the df smaller because with too many examples dspy is super slow but I realized it is not needed because later in the code I am only taking the last 25 examples anyways